language/compiler/ir-to-bytecode/src/compiler.rs

```diff
@@ -661,6 +664,32 @@ fn compile_explicit_dependency_declarations(
     Ok(())
 }
 
+fn compile_friends(
+    context: &mut Context,
+    address_opt: Option<AccountAddress>,
+    friends: Vec<ast::FriendDeclaration>,
+) -> Result<Vec<ModuleHandle>> {
+    let mut friend_decls = vec![];
+    let mut friend_decls_set = HashSet::new();
+    for friend in friends {
+        let ident = match (address_opt, friend.ident) {
+            (Some(address), ModuleIdent::Transaction(name)) => {
+                QualifiedModuleIdent { address, name }
+            }
+            (None, ModuleIdent::Transaction(name)) => bail!(
+                "Invalid friend '{}'. No address specified for script so cannot resolve friend",
```

Comment:
I'd remove this, you can just look it up in the import right?

Reply:
Not exactly. Looking it up from the import is my original design, but then I realized that all imports are eventually compiled to `module_handles`, and given the separation of the `module_handles` and `friend_decls` table, I'd prefer to have them separated.

Reply:
Yeah I don't think that really does much for the IR here. They need to be distinct in the file format, but there is no reason they need to be in the source language or even the IR

Reply:
To put this another way, the importing/aliasing doesn't exist at the file format level, it is purely a higher level construct. So it makes sense that it overlaps in places that don't overlap in the file format

Reply:
Yes, I think I am on the same mind here. In fact, they are not distinct in the source language, in a sense that we can do
```
use 0x2::A;
friend A;
```

Or even with an alias
```
use 0x2::A as B;
friend B;
```

I think I encountered some difficulties in doing the same thing on the IR compiler. In the IR compiler, there exists a much straightforward mapping from the `imports` to the `module_handles` in that every `import` will end-up in the `module_handles`, regardless of whether the imported module is used to refer to a struct or a function, or it may be possible that the imported module is not even used.

Splitting the `friends` from the `imports` would mean that, we need to peel off some modules from the `imports` pool (so they don't land in `module_handles`), and how to find those modules are not exactly very easy. (I mean, it surely can be done but kinds of breaks the straightforward mappings maintained by the IR compiler).

This problem does not really exist in the source compiler because in the source language, the `use` is indeed just for aliasing, but for IR compiler, the `import` is intended for both aliasing and marking this module as a dependency.

Reply:
Ah yeah, if the imports always get added to the module_handles work. Its fine to keep this separate for now.